qiskit_machine_learning/algorithms/distribution_learners/qgan/quantum_generator.py

```diff
@@ -417,10 +421,22 @@ def gradient_function(current_point):
                 quantum_instance, params=current_point, shots=self._shots
             )
             prediction_generated = discriminator.get_label(generated_data, detach=True)
-            op = ~CircuitStateFn(primitive=self.generator_circuit)
+            op = CircuitStateFn(primitive=self.generator_circuit)
             grad_object = gradient_object.convert(operator=op, params=free_params)
             value_dict = {free_params[i]: current_point[i] for i in range(len(free_params))}
-            analytical_gradients = np.array(grad_object.assign_parameters(value_dict).eval())
+            if quantum_instance.backend is not None:
+                grad_object = (
+                    CircuitSampler(quantum_instance.backend).convert(grad_object, value_dict).eval()
+                )
+                if quantum_instance.is_statevector:
+                    analytical_gradients = np.array(grad_object)
+                else:
+                    analytical_gradients = np.zeros((len(grad_object), grad_object[0].shape[1]))
+                    for i, _ in enumerate(grad_object):
+                        sparse_gradients = grad_object[i].tocoo()
+                        analytical_gradients[i, sparse_gradients.col] = sparse_gradients.data
```

Comment:
Could it be simpler?
```suggestion
                        analytical_gradients[i, :] = grad_object[i].todense()
```

Reply:
The reason why it was made like this was to avoid the use of any structure like `todense()` which could scale up poorly for a high number of qubits. Here I used the whole structure without having to use right after the `CircuitSampler` , `to_matrix_op()`. If the scalability using `todense()` is okay, then yes it can be simpler!

Reply:
I understand, but you collect rows from `grad_object` into `analytical_gradients` which is dense by creation! So, there's no reason to keep intermediate object as sparse objects. Correct me if I missed anything here.